### PR TITLE
[IMP] pos_loyalty: sell physical giftcard from terminal

### PR DESCRIPTION
**Pull request by taal-odoo**
Before the commit:
---
Gift card codes were generated automatically, without any option for manual input or selling physical gift cards.

After the commit:
---
- Added the "Sell Physical Gift Card" option under the "Gift Card" product in POS.
- Allows for the manual input of gift card numbers.
- Orders generate gift cards with entered numbers.
- Existing gift cards can have amounts added.

Related PR:
- Upgrade: https://github.com/odoo/upgrade/pull/6363

task-3647760

### CODE REVIEW

File: addons/pos_loyalty/static/src/utils/manage_giftcard_popup/manage_giftcard_popup.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    buttonClick(button) {
+        const lines = this.state.inputValue.split("\n").filter((line) => line !== "");
+        if (lines.includes(button.label)) {
+            this.state.inputValue = lines.filter((line) => line !== button.label).join("\n");
+            button.isSelected = false;
+        } else {
+            this.state.inputValue = lines.concat(button.label).join("\n");
+            button.isSelected = true;
+        }
+    }
```
Comment (paso-odoo) [inline addons/pos_loyalty/static/src/utils/manage_giftcard_popup/manage_giftcard_popup.js:60]:
```suggestion
```

File: addons/pos_loyalty/static/src/overrides/components/product_screen/order_summary/order_summary.js [hunk 2]
```diff
@@ -94,4 +96,73 @@ patch(OrderSummary.prototype, {
...
+                if (isNaN(points)) {
+                    console.error("Invalid amount value:", points);
+                    return;
+                }
+                code = code.trim();
+                const res = await this.pos.data.searchRead(
+                    "loyalty.card",
+                    ["&", ["program_type", "=", "gift_card"], ["code", "=", code]],
+                    [],
+                    { limit: 1 }
```
Comment (paso-odoo) [inline addons/pos_loyalty/static/src/overrides/components/product_screen/order_summary/order_summary.js:127]:
```suggestion
```

**Review by paso-odoo — CHANGES_REQUESTED** [review]

File: addons/pos_loyalty/static/src/utils/manage_giftcard_popup/manage_giftcard_popup.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        if (inputValue.trim() === "") {
+            this.state.error = true;
+            return false;
+        }
+        if (amountValue.trim() === "") {
+            this.state.amountError = true;
+            return false;
+        }
+        this.state.error = "";
+        this.state.amountError = "";
```
Comment (paso-odoo) [inline addons/pos_loyalty/static/src/utils/manage_giftcard_popup/manage_giftcard_popup.js:59]:
```suggestion
```

**Review by paso-odoo — CHANGES_REQUESTED** [review]

File: addons/pos_loyalty/static/src/overrides/components/product_screen/order_summary/order_summary.js [hunk 4]
```diff
@@ -92,4 +93,77 @@ patch(OrderSummary.prototype, {
...
+                if (isNaN(points)) {
+                    console.error("Invalid amount value:", points);
+                    return;
+                }
+                code = code.trim();
+                const res = await this.pos.data.searchRead(
+                    "loyalty.card",
+                    ["&", ["program_type", "=", "gift_card"], ["code", "=", code]],
+                    []
+                );
```
Comment (davidmonnom) [inline addons/pos_loyalty/static/src/overrides/components/product_screen/order_summary/order_summary.js:159]: Nice check

**Review by davidmonnom — CHANGES_REQUESTED** [review]
[Screencast from 07-06-24 08:51:14.webm](https://github.com/odoo/odoo/assets/63289800/9cd7deab-45c5-41e6-82b0-dc538b6047c7)

Here a little video, about an error, can you check if the user try to create two gift card with the same code ? And if yes block him.

Also, when refreshing the page, physical gift card informations are lost.

Good work!

**Comment by davidmonnom** [comment]
Hey it looks good, but can you write some tests about this new feature ?

**Comment by taal-odoo** [comment]
Hello @davidmonnom 
Thanks for the suggestions, I have pushed the changes.

**Review by davidmonnom — CHANGES_REQUESTED** [review]

File: addons/pos_loyalty/static/src/overrides/components/product_screen/order_summary/order_summary.js [hunk 5]
```diff
@@ -92,4 +94,89 @@ patch(OrderSummary.prototype, {
...
+                }
+
+                await this._updateGiftCardOrderline(code, points);
+
+                const existingGiftCard = res.length > 0 ? res[0] : null;
+                const newGiftCardCode = res.length === 0 ? code : null;
+                this.currentOrder.processGiftCard(existingGiftCard, newGiftCardCode, points);
+
+                // update indexedDB
+                this.pos.data.syncDataWithIndexedDB(this.pos.data.records);
```
[the remark this answers is not in the document]
Comment (taal-odoo) [reply addons/pos_loyalty/static/src/overrides/components/product_screen/order_summary/order_summary.js:178]: Yes, I thought so too, but it seems that changes in `uiState` do not trigger a sync to IndexedDB. 
Even before this commit, changes in `couponPointChanges` didn't call for a sync to IndexedDB. As we know the Gift Card order line gets stored. As a result, when the order line is added back after a refresh, a new `couponPointChanges` was created.

Do let me know if I have misunderstood here. Thanks.

**Comment by paan-odoo** [comment]
Hello @davidmonnom 
Changes are done.
Thanks

**Review by davidmonnom — CHANGES_REQUESTED** [review]
Hey lgtm, just a last point, can you verify the giftcard price when changing line price after adding a giftcard ?

**Review by davidmonnom — APPROVED** [review]

**Comment by caburj** [comment]
robodoo r+

**Pull request closed without merging**